src/pymor/algorithms/rand_la.py

```diff
@@ -0,0 +1,362 @@
+    
+    if range_product is None:
+        range_product = IdentityOperator(A.range)
+    else:
+        assert isinstance(range_product, Operator)
+        
+    if source_product is None:
+        source_product = IdentityOperator(A.source)
+    else:
+        assert isinstance(source_product, Operator)
+    
+    R = A.source.random(l, distribution='normal')
+
+    if iscomplex:
+        R += 1j*A.source.random(l, distribution='normal')
+
+    Q = A.apply(R)
+    gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    for i in range(q):
+        Q = A.apply_adjoint(range_product.apply(Q))
+        Q = source_product.apply_inverse(Q)
+        gram_schmidt(Q, source_product, atol=0, rtol=0, copy=False)
+        Q = A.apply(Q)
+        gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
+
+    if return_rand:
+        return Q, R
+    else:
+        return Q
+
+
+@defaults('p', 'q', 'modes')
+def random_generalized_svd(A, range_product=None, source_product=None, modes=3, p=20, q=2):
+    """Randomized SVD of a |VectorArray| or an |Operator|. 
+
+    Saibaba AK, Hart J, van BloemenWaanders B. Randomized algorithms for
+    generalized singular value decomposition with application to sensitivity analysis. Numer Linear Algebra Appl.
+    2021;28:e2364. 
+    https://doi.org/10.1002/nla.2364
+
+    Viewing the 'A' as a 'A.dim' x 'len(A)' matrix, the return value
+    of this method is the randomized singular value decomposition of 'A', where the
+    inner product on R^('dim(A)') is given by 'range_product' and the inner product on R^('len(A)')
+    is given by 'source_product'. 
+        
+        A = U*\Sigma*V*source_product
+    
+    Parameters
+    ----------
+    A : 
+        The |VectorArray| or |Operator| for which the randomized SVD is to be computed. 
+    range_product : 
+        Range product |Operator| w.r.t which the randomized SVD is computed
+    source_product :
+        Source product |Operator| w.r.t which the randomized SVD is computed
+    modes : 
+        The first 'modes' approximated singular values and vectors are returned. 
+    p :
+        If not '0', adds 'Oversampling' colums to the randomly sampled matrix. 
+    q : 
+        If not '0', performs 'q' so called power iterations to increase the relative weight
+        of the first singular values.
+
+    Returns
+    -------
+    U 
+        |VectorArray| of approximated left singular vectors 
+    s 
+        One-dimensional |NumPy array| of the approximated singular values
+    V
+        |VectorArray| of the approximated right singular vectors
+    """
+    logger = getLogger('pymor.algorithms.rand_la')
+    
+    assert isinstance(A, VectorArray) or isinstance(A, Operator)
+    
+    if isinstance(A, VectorArray):
+        A = VectorArrayOperator(A)
+   
+    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
+    assert p >= 0 and isinstance(p, int)
+    assert q >= 0 and isinstance(q, int)
```

Comment:
I did some quick tests and I can create some errors down the line with small operators. It should be checked that oversampling does not exceed the rank of the operator.
Maybe like so:
```suggestion
    assert 0 <= modes <= max(A.source.dim, A.range.dim) and isinstance(modes, int)
    assert 0 <= p <= max(A.source.dim, A.range.dim) - modes and isinstance(p, int)
    assert q >= 0 and isinstance(q, int)
```

Alternatively, the oversampling parameter could be capped.

Reply:
It would be great if you can remember/save the setup that forced the errors so we can use that as a test case.

Reply:
I simply used `NumpyMatrixOperator(np.random.rand(5, 5))`. The default arguments already lead to errors since `5 < p + q = 23`